src/Workspaces/Core/Portable/Workspace/Workspace_EventsLegacy.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+
+namespace Microsoft.CodeAnalysis;
+
+public abstract partial class Workspace
+{
+    // Allows conversion of legacy event handlers to the new event system.
+    private readonly ConcurrentDictionary<(object, string), IDisposable> _disposableEventHandlers = new();
```

Comment:
what sit he 'object/string' here?

Reply:
It was the easiest way for me to get a EventHandler<TEventArgs> that can be used across the (Add/Remove)EventHandler code to map to the proper disposer.

Reply:
I'm considering changing the "string" part of it to an enum that would represent the string event types represented in Workspace_Events.cs. If the EventMap is tied to the workspace now, it doesn't seem to need to be as general as it was probably required before.

Reply:
Went ahead and made that change to use a strong typed enum for the event type rather than a string.